pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -608,172 +721,153 @@ protected SegmentDataManager unregisterSegment(String segmentName) {
-      _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.SEGMENT_DOWNLOAD_FROM_REMOTE_FAILURES, 1L);
-      if (_peerDownloadScheme == null) {
-        throw e;
+      File untarredSegmentDir;
+      if (_isStreamSegmentDownloadUntar && zkMetadata.getCrypterName() == null) {
+        _logger.info("Downloading segment: {} using streamed download-untar with maxStreamRateInByte: {}", segmentName,
+            _streamSegmentDownloadUntarRateLimitBytesPerSec);
+        AtomicInteger attempts = new AtomicInteger(0);
+        try {
+          untarredSegmentDir = SegmentFetcherFactory.fetchAndStreamUntarToLocal(downloadUrl, tempRootDir,
+              _streamSegmentDownloadUntarRateLimitBytesPerSec, attempts);
+          _logger.info("Downloaded and untarred segment: {} from: {}, attempts: {}", segmentName, downloadUrl,
+              attempts.get());
+        } finally {
+          _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.SEGMENT_STREAMED_DOWNLOAD_UNTAR_FAILURES,
+              attempts.get());
+        }
+      } else {
+        File segmentTarFile = new File(tempRootDir, segmentName + TarGzCompressionUtils.TAR_GZ_FILE_EXTENSION);
+        SegmentFetcherFactory.fetchAndDecryptSegmentToLocal(downloadUrl, segmentTarFile, zkMetadata.getCrypterName());
+        _logger.info("Downloaded tarred segment: {} from: {} to: {}, file length: {}", segmentName, downloadUrl,
+            segmentTarFile, segmentTarFile.length());
+        untarredSegmentDir = untarSegment(segmentName, segmentTarFile, tempRootDir);
       }
-      downloadFromPeersWithoutStreaming(segmentName, zkMetadata, tarFile);
-      downloadSuccess = true;
-      return tarFile;
+      File indexDir = moveSegment(segmentName, untarredSegmentDir);
+      _logger.info("Downloaded segment: {} from: {} to: {}", segmentName, downloadUrl, indexDir);
+      return indexDir;
+    } catch (Exception e) {
+      _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.SEGMENT_DOWNLOAD_FROM_REMOTE_FAILURES, 1);
+      throw e;
     } finally {
-      if (!downloadSuccess) {
-        _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.SEGMENT_DOWNLOAD_FAILURES, 1L);
-      }
       if (_segmentDownloadSemaphore != null) {
         _segmentDownloadSemaphore.release();
       }
+      FileUtils.deleteQuietly(tempRootDir);
     }
   }
 
-  protected void downloadFromPeersWithoutStreaming(String segmentName, SegmentZKMetadata zkMetadata, File destTarFile)
+  @VisibleForTesting
+  File downloadSegmentFromPeers(SegmentZKMetadata zkMetadata)
       throws Exception {
+    String segmentName = zkMetadata.getSegmentName();
     Preconditions.checkState(_peerDownloadScheme != null, "Peer download is not enabled for table: %s",
         _tableNameWithType);
+    _logger.info("Downloading segment: {} from peers", segmentName);
+    File tempRootDir = getTmpSegmentDataDir("tmp-" + segmentName + "-" + UUID.randomUUID());
```

Comment:
Good point! Seems we no longer expose them in test

Reply:
Actually after a second thought, I feel keeping them within the method might be safer. In case download segment from deep store failed, we want to delete the tmp dir before trying to download from peers